src/ray/util/event_test.cc

```diff
@@ -0,0 +1,113 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fstream>
+#include <set>
+#include <thread>
+#include "gtest/gtest.h"
+#include "ray/util/event.h"
+
+namespace ray {
+
+class TestEventReporter : public LogBasedEventReporter {
+ public:
+  static std::vector<rpc::Event> event_list;
+  virtual void Init() override {}
+  virtual void Report(rpc::Event &event) override { event_list.push_back(event); }
+  virtual void Close() override {}
+  virtual ~TestEventReporter() {}
+  virtual std::string GetReporterKey() override { return "test.event.reporter"; }
+
+ protected:
+  virtual std::string EventToString(rpc::Event &event) override { return ""; }
+};
+
+std::vector<rpc::Event> TestEventReporter::event_list = std::vector<rpc::Event>();
+
+void CheckEventDetail(rpc::Event &event, std::string job_id, std::string node_id,
+                      std::string task_id, std::string source_type, std::string severity,
+                      std::string label, std::string message) {
+  int custom_key_num = 0;
+  auto mp = (*event.mutable_custom_field());
+
+  if (job_id != "") {
+    EXPECT_EQ(mp["job_id"], job_id);
+    custom_key_num++;
+  }
+  if (node_id != "") {
+    EXPECT_EQ(mp["node_id"], node_id);
+    custom_key_num++;
+  }
+  if (task_id != "") {
+    EXPECT_EQ(mp["task_id"], task_id);
+    custom_key_num++;
+  }
+  EXPECT_EQ(mp.size(), custom_key_num);
+  EXPECT_EQ(rpc::Event_SourceType_Name(event.source_type()), source_type);
+  EXPECT_EQ(rpc::Event_Severity_Name(event.severity()), severity);
+  EXPECT_EQ(event.label(), label);
+  EXPECT_EQ(event.message(), message);
+  EXPECT_EQ(event.source_pid(), getpid());
+}
+
+TEST(EVENT_TEST, TEST_BASIC) {
+  ray::EventManager::Instance().ClearReporters();
```

Comment:
Can you also verify if `Publish` has been called?

Reply:
Of course, the AddReporters, ClearPeporters (along with the Pushlish function) would not be thread-safe. But we assume default initialization and shutdown are placed in the construction and destruction of a resident class, or at the beginning and end of a process. 

Do you mean we have to add mutex to prevent these problems?

Reply:
Oh, no. I just meant to have a check to make sure `EventManager::Instance().Publish(event);` is called whenever `RAY_EVENT()` is called. 

Also about thread-safety. Can you mention them somewhere with NOTE?